### PR TITLE
Adds parallel map-reduce recipe

### LINKED ISSUE #329: Add "Parallel map reduce" example
Showcase a basic map-reduce scheme with [rayon](https://crates.io/crates/rayon) crate.

Please discuss the idea here before starting implementation.
cc https://github.com/rust-lang-nursery/rust-cookbook/issues/272

### PR DESCRIPTION
**Pull request by j-haj**
fixes #329

* Adds example in `concurrency.md`
* Adds links to `intro.md`

### CODE REVIEW

File: src/concurrency.md [hunk 1]
```diff
@@ -73,6 +74,63 @@ fn main() {
...
+        Person { age: 17 },
+        Person { age: 31 },
+        Person { age: 30 },
+    ];
+
+    let gt_30 = |x: u32| x > 30;
+    let num_over_30 = v.par_iter().filter(|&ref x| gt_30(x.age)).count();
+    let sum_over_30 = v.par_iter()
+        .filter(|&ref x| gt_30(x.age))
+        .map(|ref x| x.age)
```
Comment (budziq) [inline src/concurrency.md]: I'd suggest mapping first and then using `filter(gt_30)` instead of closure in closure

File: src/concurrency.md [hunk 2]
```diff
@@ -73,6 +74,63 @@ fn main() {
...
+        .reduce(|| 0, |x, y| x + y);
+    
+    // Rayon also has a sum() method, which yields the same results as the
+    // reduction on the previous line
```
Comment (budziq) [inline src/concurrency.md]: I'd move this comment to the textual description above

File: src/concurrency.md [hunk 3]
```diff
@@ -73,6 +74,63 @@ fn main() {
...
+    let alt_sum_30: u32 = v.par_iter()
+        .filter(|&ref x| gt_30(x.age))
+        .map(|ref x| x.age)
+        .sum();
+
+    let avg_over_30 = sum_over_30 as f32 / num_over_30 as f32;
+    let alt_avg_over_30 = alt_sum_30 as f32 / num_over_30 as f32;
+
+    assert!((avg_over_30 - alt_avg_over_30).abs() < 1e-8);
```
Comment (budziq) [inline src/concurrency.md]: I'd suggest using `std::f32::EPSILON` instead of `1e-8`

File: src/concurrency.md [hunk 4]
```diff
@@ -73,6 +74,63 @@ fn main() {
...
+        .reduce(|| 0, |x, y| x + y);
+    
+    // Rayon also has a sum() method, which yields the same results as the
+    // reduction on the previous line
+    let alt_sum_30: u32 = v.par_iter()
+        .filter(|&ref x| gt_30(x.age))
+        .map(|ref x| x.age)
+        .sum();
+
+    let avg_over_30 = sum_over_30 as f32 / num_over_30 as f32;
```
Comment (budziq) [inline src/concurrency.md]: I'd probably move casts to above and store the variables as `f32` but it's a matter of taste.

File: src/concurrency.md [hunk 5]
```diff
@@ -73,6 +74,63 @@ fn main() {
...
+        Person { age: 19 },
+        Person { age: 42 },
+        Person { age: 17 },
+        Person { age: 17 },
+        Person { age: 31 },
+        Person { age: 30 },
+    ];
+
+    let gt_30 = |x: u32| x > 30;
+    let num_over_30 = v.par_iter().filter(|&ref x| gt_30(x.age)).count();
```
Comment (budziq) [inline src/concurrency.md]: the `&ref` and `ref` are not needed anywhere in this example

File: src/concurrency.md (hunk 4 above)
Comment (j-haj) [reply src/concurrency.md]: I run into a type inference issue with the `alt_sum_30` variable if I moved `as f32` to the line where `alt_sum_30` is initialized. So far the easiest way to get rid of that error is to make `alt_sum_30` a `u32` and then convert to `f32` later. Any suggestions?

Comment (budziq) [reply src/concurrency.md]: right `sum` is a template so type inference gets wonky here. 

```rust
 let alt_sum_30: f32 = v.par_iter()
        .filter(|x| gt_30(x.age))
        .map(|x| x.age as f32)
        .sum();
```
```rust
    let alt_sum_30 = v.par_iter()
        .filter(|x| gt_30(x.age))
        .map(|x| x.age as f32)
        .sum::<f32>();
```

or the cleanest solution, leaving is as it was, summed into u32 and casting later on ;)

**Comment by j-haj** [comment]
Thanks for all the comments! I have pushed an updated version

**Comment by budziq** [comment]
Nicely done!

**Pull request merged**